lib/vrl/compiler/src/type_def.rs

```diff
@@ -926,6 +952,197 @@ impl TypeDef {
         self
     }
 
+    fn update_segment<'a, I>(
+        kind: &KindInfo,
+        mut path: std::iter::Peekable<I>,
+        newkind: &KindInfo,
+    ) -> KindInfo
+    where
+        I: std::iter::Iterator<Item = &'a SegmentBuf> + Clone,
+    {
+        match kind {
+            KindInfo::Unknown => {
+                // The kind is unknown and thus there is nothing to remove.
+                kind.clone()
+            }
+            KindInfo::Known(kinds) => KindInfo::Known(
+                kinds
+                    .iter()
```

Comment:
Can we make this `into_iter`, to avoid cloning inner types down below? Nothing blocking, but might be worth sneaking in that improvement.

Reply:
Since `kind` is a reference, so is `kinds` is also a reference, which means `into_iter` still creates an iterator over references.

I did toy with the idea of making `update_segment` take an owned `kind` which would make this possible, but would require the caller to clone. I think this way is more efficient, albeit only slightly.

I did also play with mutating the object directly, but that started turning into a Phd thesis.

Reply:
Ah yes, I missed that `kind` was a reference. Fair enough 👍